pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -144,43 +146,46 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (reconcile
 
 	// If we do not have all of the needed BMC credentials, set our
 	// operational status to indicate missing information.
-	if instance.Spec.BMC.IP == "" || instance.Spec.BMC.Username == "" || instance.Spec.BMC.Password == "" {
-		if instance.SetErrorMessage(MissingBMCCredentialsMsg) {
-			reqLogger.Info(
-				"adding error message",
-				"message", MissingBMCCredentialsMsg,
-			)
-			if err := r.saveStatus(instance); err != nil {
-				reqLogger.Error(err, "failed to update error message")
-				return reconcile.Result{}, err
-			}
-		}
-		if instance.SetOperationalStatus(metalkubev1alpha1.OperationalStatusError) {
-			reqLogger.Info(
-				"setting operational status",
-				"newStatus", metalkubev1alpha1.OperationalStatusError,
-			)
-			if err := r.client.Update(context.TODO(), instance); err != nil {
-				reqLogger.Error(err, "failed to update operational status")
-				return reconcile.Result{}, err
-			}
-			return reconcile.Result{Requeue: true}, nil
-		}
+	if instance.Spec.BMC.IP == "" {
+		reqLogger.Info(bmc.MissingIPMsg)
+		err := r.setErrorCondition(request, instance, bmc.MissingIPMsg)
+		// Without the BMC IP there's no more we can do, so we're
+		// going to return the emtpy Result anyway, and don't need to
+		// check err.
+		return reconcile.Result{}, err
+	}
+
+	if instance.Spec.BMC.Credentials == nil {
+		reqLogger.Info(bmc.MissingCredentialsMsg)
+		err := r.setErrorCondition(request, instance, bmc.MissingCredentialsMsg)
+		// Without the BMC IP there's no more we can do, so we're
+		// going to return the emtpy Result anyway, and don't need to
+		// check err.
+		return reconcile.Result{}, err
 	} else {
-		if instance.SetErrorMessage("") {
-			reqLogger.Info("clearing error message")
-			if err := r.saveStatus(instance); err != nil {
-				reqLogger.Error(err, "failed to clear error message")
-				return reconcile.Result{}, err
-			}
+
+		bmcCreds, err := r.getBMCCredentials(request, instance)
+		if err != nil {
+			// If we can't fetch the BMC settings there's no more we can do.
+			reqLogger.Info("missing BMC Credentials")
+			return reconcile.Result{}, err
 		}
-		newOpStatus := metalkubev1alpha1.OperationalStatusOffline
-		if instance.Spec.Online {
-			newOpStatus = metalkubev1alpha1.OperationalStatusOnline
+		validCreds, reason := bmcCreds.AreValid()
+		if !validCreds {
+			reqLogger.Info("invalid BMC Credentials", "reason", reason)
+			err := r.setErrorCondition(request, instance, reason)
+			// We won't see a change to the secret, so put ourselves
+			// back into the queue after a brief wait.
```

Comment:
We could update to watch Secrets, too, but this seems fine for now

Reply:
Yeah, I chatted with Michael about how to do that and am going to experiment with it next week.